ops/_main.py

```diff
@@ -515,19 +499,48 @@ def _emit(self):
             self.framework.reemit()
 
         # Emit the Juju event.
-        _emit_charm_event(self.charm, self.dispatcher.event_name, self._juju_context)
+        self._emit_charm_event(self.dispatcher.event_name)
         # Emit collect-status events.
         ops.charm._evaluate_status(self.charm)
 
+    def _get_event_to_emit(self, event_name: str) -> Optional[ops.framework.BoundEvent]:
+        try:
+            return getattr(self.charm.on, event_name)
+        except AttributeError:
+            logger.debug('Event %s not defined for %s.', event_name, self.charm)
+        # If the event is not defined, return None.
+
+    def _emit_charm_event(self, event_name: str):
+        """Emits a charm event based on a Juju event name.
+
+        Args:
+            charm: A charm instance to emit an event from.
+            event_name: A Juju event name to emit on a charm.
+            juju_context: An instance of the _JujuContext class.
+        """
+        event_to_emit = self._get_event_to_emit(event_name)
+
+        # If the event is not supported by the charm implementation, do
+        # not error out or try to emit it. This is to support rollbacks.
+        if event_to_emit is not None:
+            args, kwargs = _get_event_args(self.charm, event_to_emit, self._juju_context)
+            logger.debug('Emitting Juju event %s.', event_name)
+            event_to_emit.emit(*args, **kwargs)
```

Comment:
Minor style change that I think is easier to follow, up to you
```suggestion
        if event_to_emit is None:
            # If the event is not supported by the charm implementation, do
            # not error out or try to emit it. This is to support rollbacks.
            return
        args, kwargs = _get_event_args(self.charm, event_to_emit, self._juju_context)
        logger.debug('Emitting Juju event %s.', event_name)
        event_to_emit.emit(*args, **kwargs)
```

Reply:
This was code that was just moved, not added in this PR, so my inclination was to leave it alone rather than changing *and* moving it. But I agree it is nicer that way, so will do.